### PR TITLE
UP-4096:  Convert sitemap to a portlet

### PR DESCRIPTION
**Pull request by cbeach47**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] [message properties][] have been updated with new phrases
-   [x] view conforms with [WCAG 2.0 AA][]

##### Description of change

XSL sitemap converted to a portlet.  This sitemap portlet builds the sitemap from the v4.3 layout.json API call.

PR is almost done - not sure the best way to find the content path in building the tab and portlet links.

https://issues.jasig.org/browse/UP-4096

Should be merged with the uPortal-start PR:  https://github.com/Jasig/uPortal-start/pull/93

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		// Begin tab row
+       	var tabRowTemplate = document.getElementById('sitemap-tab-row-template');
+		var tabRow = document.importNode(tabRowTemplate.content, true);
+		var tabCounter = 0;
+		_.forEach(response.layout.navigation.tabs, function (tab, tabIndex) {
+			sitemapJsonCheck(tab, ['name','ID','content'], "Missing required object path [layout.navigation.tabs] > ");
+			tabCounter++;
+			// Setup tab link
+			var tabTemplate = document.getElementById('sitemap-tab-template');
+           	var tabHeader = document.importNode(tabTemplate.content, true);
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: Should lines 99 and 100 be at the same indent level?

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		
+			_.forEach(tab.content, function (parentContent, parentContentIndex) {
+				sitemapJsonCheck(parentContent, ['content'], "Missing required object path [layout.navigation.tabs] > content > ");
+
+				_.forEach(parentContent.content, function (portlet, portletIndex) {
+					sitemapJsonCheck(portlet, ['name', 'fname', 'ID'], "Missing required object path [layout.navigation.tabs] > content > content > ");
+
+					// Setup portlet link
+					var portletTemplate = document.getElementById('sitemap-tab-portlet-template');
+           			var portletListItem = document.importNode(portletTemplate.content, true);
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: Should line 114 and 115 be at the same indent level?

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+				// Add tab row to page, and initialize a new tab row
+				document.getElementById('sitemap-holder').appendChild(tabRow);
+				tabRow = document.importNode(tabRowTemplate.content, true);
+			}
+		});
+    })
+    // Let user know and log error to browser console
+    .catch(function(error) {
+        var errTemplate = document.getElementById('sitemap-tab-row-template');
+		var errDiv = document.importNode(errTemplate.content, true);
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: Should line 141 and 142 be at the same indent level?

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		errDiv.textContent = '${i18n_error_loading_sitemap}';	
+		document.getElementById('sitemap-holder').appendChild(errDiv);
+        console.log('${i18n_error_loading_sitemap}', error)
+    });
+
+// If a path check fails, it'll throw an error.
+function sitemapJsonCheck(jsonObj, pathChecks, errMsg) {
+	_.forEach(pathChecks, function(pathCheck, errIndex) {
+		if(!_.has(jsonObj, pathCheck)) {
+			var error = new Error(errMsg+pathCeck);
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: spacing around infix operator?

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+					sitemapJsonCheck(portlet, ['name', 'fname', 'ID'], "Missing required object path [layout.navigation.tabs] > content > content > ");
+
+					// Setup portlet link
+					var portletTemplate = document.getElementById('sitemap-tab-portlet-template');
+           			var portletListItem = document.importNode(portletTemplate.content, true);
+           			// Add content to portlet template
+           			var portletTitle = portletListItem.querySelector('span');
+           			portletTitle.textContent = portlet.name;
+					var portletLink = portletListItem.querySelector('a');
+           			portletLink.href = '/uPortal/f/' + tab.ID + '/p/'+portlet.fname+'.'+portlet.ID+'/max/render.uP';
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: spaces around infix operators?

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+           			// Add portlet to tab list
+           			portletList.appendChild(portletListItem);  
+				});
+			});
+
+			tabRow.appendChild(tabHeader);
+			
+			if(tabIndex == (response.layout.globals.tabsInTabGroup-1)) {
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: :art: prefer type safe equality `===`

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp (hunk 5 above)
Comment (cbeach47) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: I think there's a better way to setup this link.

Comment (ChristianMurphy) [reply uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: @drewwills are there some JSP/Java helper variables/functions that could help generate this link?

Comment (drewwills) [reply uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: I sent something to @cbeach47 in chat.

File: uPortal-webapp/src/main/resources/layout/theme/respondr/regions.xsl [hunk 7]
```diff
@@ -292,19 +292,14 @@
     <!-- ========== TEMPLATE: FOOTER FIRST REGION =============== -->
     <!-- ======================================================== -->
     <!--
-     | This template renders region intended to hold the site navigation.
+     | This template renders the region intended to hold the sitemap.
```
Comment (drewwills) [inline uPortal-webapp/src/main/resources/layout/theme/respondr/regions.xsl]: We should remove this comment;  some humans may have this intention, but the intention is not inherent in the theme.

File: uPortal-webapp/src/main/resources/layout/theme/respondr/regions.xsl [hunk 8]
```diff
@@ -314,8 +309,7 @@
     <!-- ========== TEMPLATE: FOOTER SECOND REGION (License links) =============== -->
     <!-- ========================================================================= -->
     <!--
-     | This template renders region intended to hold the license portlet.
-     | TODO:  Move footer.nav to footer.first and convert to a portlet (see UP-4103)
+     | This template renders the region intended to hold the license portlet.
```
Comment (drewwills) [inline uPortal-webapp/src/main/resources/layout/theme/respondr/regions.xsl]: Same as above.

File: uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+--%>
+<%@ include file="/WEB-INF/jsp/include.jsp"%>
+
+<div class="container-fluid" id="sitemap-holder">
```
Comment (drewwills) [inline uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp]: Indentation looks off in this file.

File: uPortal-webapp/src/main/webapp/media/skins/respondr/common/less/footer.less [hunk 10]
```diff
@@ -20,6 +20,11 @@
 /* ==========================================================================
    Footer Nav
    ========================================================================== */
+#sitemap-tab-header {
```
Comment (drewwills) [inline uPortal-webapp/src/main/webapp/media/skins/respondr/common/less/footer.less]: There's probably a better place for this style rule.  It's portlet-specific, so should probably go into `skins/respondr/common/less/views`.

**Comment by drewwills** [comment]
I implemented all the suggestions on the PR.  I think this item is ready now.  Let's aim to merge after we have a quick look tomorrow.

**Comment by drewwills** [comment]
Some changes I made yesterday...

  - There was a `<template>` that contained a `<div>` with an id of `sitemap-tab-header`, which (after JS execution) was producing multiple elements with the same id on the page;  I refactored to a class of `up-sitemap-tab-header`.  (I prefer it when all custom CSS classes we define have a `up-` prefex, or perhaps `ann-` in something like the Announcements portlet.)
  - I moved the custom LESS code for this portlet into the `views/` dir, which exists to separate portlet-specific style rules from (more general) theme-level style rules.
  - For me, the `tabHeader` (tab) nodes were being added at the wrong level of the DOM;  I addressed that.

**Comment by ChristianMurphy** [comment]
CI is flagging:

```
/home/travis/build/Jasig/uPortal/uPortal-webapp/src/test/java/org/apereo/portal/portlets/sitemap/SitemapTest.java:82: error: cannot find symbol
            transformer.setParameter(SitemapPortletController.USE_TAB_GROUPS, useTabGroups);
                                     ^
  symbol:   variable SitemapPortletController
  location: class SitemapTest
/home/travis/build/Jasig/uPortal/uPortal-webapp/src/test/java/org/apereo/portal/portlets/sitemap/SitemapTest.java:83: error: cannot find symbol
            transformer.setParameter(SitemapPortletController.USER_LANG, "en_US");
                                     ^
  symbol:   variable SitemapPortletController
  location: class SitemapTest
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
2 errors
```

**Pull request merged**